src/CPUSolver.cpp

```diff
@@ -355,111 +262,184 @@ void CPUSolver::normalizeFluxes() {
+                               _old_scalar_flux(r,e), 2);
+      }
     }
+  }
+
+  else if (res_type == FISSION_SOURCE) {
+
+    norm = _num_fissionable_FSRs * _num_groups;
+
+    double new_fission_source, old_fission_source;
+    FP_PRECISION* nu_sigma_f;
+    Material* material;
 
-    /* Compute the norm of residual of the source in the FSR */
-    if (fsr_fission_source > 0.0)
-      _source_residuals[r] = pow((fsr_fission_source - _old_fission_sources[r])
-                                 / fsr_fission_source, 2);
+    for (int r=0; r < _num_FSRs; r++) {
+      new_fission_source = 0.;
+      old_fission_source = 0.;
+      material = _FSR_materials[r];
+
+      if (material->isFissionable()) {
+        nu_sigma_f = material->getNuSigmaF();
+
+        for (int e=0; e < _num_groups; e++) {
+          new_fission_source += _scalar_flux(r,e) * nu_sigma_f[e];
+          old_fission_source += _old_scalar_flux(r,e) * nu_sigma_f[e];
+        }
 
-    /* Update the old source */
-    _old_fission_sources[r] = fsr_fission_source;
+        if (old_fission_source > 0.)
+          residuals[r] = pow((new_fission_source -  old_fission_source) /
+                              old_fission_source, 2);
+      }
+    }
   }
 
-  /* Sum up the residuals from each FSR */
-  source_residual = pairwise_sum<FP_PRECISION>(_source_residuals, _num_FSRs);
-  source_residual = sqrt(source_residual \
-                         / (_num_fissionable_FSRs * _num_groups));
+  else if (res_type == TOTAL_SOURCE) {
+
+    norm = _num_FSRs * _num_groups;
+
+    double new_total_source, old_total_source; 
+    FP_PRECISION inverse_k_eff = 1.0 / _k_eff;
+    FP_PRECISION* nu_sigma_f;
+    Material* material;
 
-  return source_residual;
+    for (int r=0; r < _num_FSRs; r++) {
+      new_total_source = 0.;
+      old_total_source = 0.;
+      material = _FSR_materials[r];
+
+      if (material->isFissionable()) {
+        nu_sigma_f = material->getNuSigmaF();
+ 
+        for (int e=0; e < _num_groups; e++) {
+          new_total_source += _scalar_flux(r,e) * nu_sigma_f[e];
+          old_total_source += _old_scalar_flux(r,e) * nu_sigma_f[e];
+        }
+
+        new_total_source *= inverse_k_eff;
+        old_total_source *= inverse_k_eff;
+      }
+
+      /* Compute total scattering source for group G */
+      for (int G=0; G < _num_groups; G++) {
+        for (int g=0; g < _num_groups; g++) {
+          new_total_source += material->getSigmaSByGroupInline(g,G)
+                              * _scalar_flux(r,g);
+          old_total_source += material->getSigmaSByGroupInline(g,G)
+                              * _old_scalar_flux(r,g);
+        }
+      }
+
+      if (old_total_source > 0.)
+        residuals[r] = pow((new_total_source -  old_total_source) / 
+                            old_total_source, 2);
+    }
+  }
+
+  /* Sum up the residuals from each FSR and normalize */
+  residual = pairwise_sum<double>(residuals, _num_FSRs);
+  residual = sqrt(residual / norm);
```

Comment:
I would recommend having `norm` set to the number of FSRs with an `old_total_source` greater than 0.